src/mr/coordinator.go

```diff
@@ -1,40 +1,199 @@
 package mr
 
-import "log"
-import "net"
-import "os"
-import "net/rpc"
-import "net/http"
-
+import (
+	"fmt"
+
+	"log"
+	"net"
+	"net/http"
+	"net/rpc"
+	"strconv"
+	"sync"
+	"sync/atomic"
+)
 
 type Coordinator struct {
-	// Your definitions here.
+	inputfiles      []string
+	reduceTasksCnt  int
+	workerCnt       int32
+	workerStatusMap SafeMap
+	// key: string representation of task
+	// value: task status, 1-idle, 2-in progress, 3-completed
+	taskStatusMap           SafeStatusMap
+	idleMapTasksQueue       chan Task
+	idleReduceTaskQueue     chan Task
+	taskInitializationMutex sync.Mutex
+	// atomic interger
+	remainingMapTasksCnt    int32
+	remainingReduceTasksCnt int32
+}
+
+type Task struct {
+	TaskType   int // 1-map, 2-reduce
+	Inputfiles []string
+	TaskId     int
+}
+
+func (t Task) toString() string {
+	return fmt.Sprintf("%v", t)
+}
+
+type SafeMap struct {
+	m  map[string]bool
+	mu sync.Mutex
+}
+
+type SafeStatusMap struct {
+	m  map[string]int
+	mu sync.Mutex
+}
+
+// gives each input file to a map task
+func (c *Coordinator) initializeMapTasks() {
+	atomic.StoreInt32(&c.remainingMapTasksCnt, int32(len(c.inputfiles)))
+	for i, file := range c.inputfiles {
+		t := Task{TaskType: 1, Inputfiles: []string{file}, TaskId: i}
+		log.Printf("Initializing map task: %v\n", t)
+		c.taskStatusMap.mu.Lock()
+		c.taskStatusMap.m[t.toString()] = 1
+		c.taskStatusMap.mu.Unlock()
+		// if queue is full, it will block until there is some consumer complete some tasks
+		c.idleMapTasksQueue <- t
+	}
+	log.Printf("IdleMapTaskQueue Len: %d\n", len(c.idleMapTasksQueue))
+
+}
 
+func (c *Coordinator) checkTaskStatus(t Task) (int, bool) {
+	c.taskStatusMap.mu.Lock()
+	defer c.taskStatusMap.mu.Unlock()
+	status, ok := c.taskStatusMap.m[t.toString()]
+	return status, ok
 }
 
 // Your code here -- RPC handlers for the worker to call.
 
-//
-// an example RPC handler.
-//
-// the RPC argument and reply types are defined in rpc.go.
-//
-func (c *Coordinator) Example(args *ExampleArgs, reply *ExampleReply) error {
-	reply.Y = args.X + 1
+func (c *Coordinator) RegisterWorker(args *RegisterWorkerArgs, reply *RegisterWorkerReply) error {
```

Comment:
RegisterWorker RPC may not be needed, since we really only care about the task status in this lab.